### ISSUE TITLE
crossentropy causes and error

### ISSUE THREAD

**Issue by OTapio** [FluxML/Flux.jl#1105]
My system and packages are

Windows 10
Julia 1.4
Flux v0.10.3
CuArrays v2.0.1

when I run the following code

```
using Flux, CuArrays
using Flux: crossentropy
model = Chain(
        Dense(728, 128, σ),
        LSTM(128, 256),
        LSTM(256, 128),
        Dense(128, 10),
        softmax
        ) |> gpu
data = [rand(728) for i in 1:100];
out  = [rand(10) for i in 1:100];

loss(x, y) = crossentropy(model(x), y);

Flux.train!(loss, params(model), zip(gpu.(data), gpu.(out)), ADAM())
```

console gives me warning
```
julia> 
┌ Warning: calls to Base intrinsics might be GPU incompatible
│   exception =
│    You called log(x::Float32) in Base.Math at special/log.jl:290, maybe you intended to call log(x::Float32) in CUDAnative at C:\Users\OTapio\.julia\packages\CUDAnative\cnQli\src\device\cuda\math.jl:73 instead?
│    Stacktrace:
│     [1] log at special/log.jl:290
│     [2] #26 at C:\Users\OTapio\.julia\packages\GPUArrays\QDGmr\src\host\broadcast.jl:64
└ @ CUDAnative C:\Users\OTapio\.julia\packages\CUDAnative\cnQli\src\compiler\irgen.jl:266
```
 and the Flux.train! gives error
```
GPU compilation of broadcast(CuArrays.CuKernelContext, CUDAnative.CuDeviceArray{Tuple{Float32,Zygote.var"#963#back#398"{Zygote.var"#396#397"{Float32}}},1,CUDAnative.AS.Global}, Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64}},Zygote.var"#1666#1673"{Zygote.Context,typeof(log)},Tuple{Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,1,CUDAnative.AS.Global},Tuple{Bool},Tuple{Int64}}}}) failed
KernelError: passing and using non-bitstype argument

Argument 4 to your kernel function is of type Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64}},Zygote.var"#1666#1673"{Zygote.Context,typeof(log)},Tuple{Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,1,CUDAnative.AS.Global},Tuple{Bool},Tuple{Int64}}}}.
That type is not isbits, and such arguments are only allowed when they are unused by the kernel.  .f is of type Zygote.var"#1666#1673"{Zygote.Context,typeof(log)} which is not isbits.
    .__context__ is of type Zygote.Context which is not isbits.
      .cache is of type Union{Nothing, IdDict{Any,Any}} which is not isbits.


check_invocation(::CUDAnative.CompilerJob, ::LLVM.Function) at validation.jl:72
macro expansion at driver.jl:185 [inlined]
macro expansion at TimerOutput.jl:228 [inlined]
codegen(::Symbol, ::CUDAnative.CompilerJob; libraries::Bool, dynamic_parallelism::Bool, optimize::Bool, strip::Bool, strict::Bool) at driver.jl:184
codegen at driver.jl:55 [inlined]
compile(::Symbol, ::CUDAnative.CompilerJob; libraries::Bool, dynamic_parallelism::Bool, optimize::Bool, strip::Bool, strict::Bool) at driver.jl:45
compile at driver.jl:41 [inlined]
#compile#175 at driver.jl:33 [inlined]
compile at driver.jl:33 [inlined]
compile at driver.jl:33 [inlined]
cufunction_slow(::Function, ::Type{T} where T, ::Int64; name::String, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at execution.jl:326
cufunction_slow at execution.jl:320 [inlined]
#223 at execution.jl:393 [inlined]
get!(::CUDAnative.var"#223#224"{String,Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},GPUArrays.var"#26#27",DataType,Int64}, ::Dict{UInt64,CUDAnative.HostKernel}, ::UInt64) at dict.jl:452
macro expansion at execution.jl:392 [inlined]
macro expansion at lock.jl:183 [inlined]
cufunction_fast(::Function, ::Type{T} where T, ::Int64; name::String, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at execution.jl:391
(::CUDAnative.var"#cufunction_fast##kw")(::NamedTuple{(:name,),Tuple{String}}, ::typeof(CUDAnative.cufunction_fast), ::Function, ::Type{T} where T, ::Int64) at execution.jl:380
getproperty at Base.jl:33 [inlined]
merge at namedtuple.jl:235 [inlined]
cufunction(::GPUArrays.var"#26#27", ::Type{Tuple{CuArrays.CuKernelContext,CUDAnative.CuDeviceArray{Tuple{Float32,Zygote.var"#963#back#398"{Zygote.var"#396#397"{Float32}}},1,CUDAnative.AS.Global},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64}},Zygote.var"#1666#1673"{Zygote.Context,typeof(log)},Tuple{Base.Broadcast.Extruded{CUDAnative.CuDeviceArray{Float32,1,CUDAnative.AS.Global},Tuple{Bool},Tuple{Int64}}}}}}; kwargs::Base.Iterators.Pairs{Symbol,String,Tuple{Symbol},NamedTuple{(:name,),Tuple{String}}}) at execution.jl:0
(::CUDAnative.var"#cufunction##kw")(::NamedTuple{(:name,),Tuple{String}}, ::typeof(CUDAnative.cufunction), ::Function, ::Type{T} where T) at execution.jl:422
macro expansion at execution.jl:157 [inlined]
gpu_call(::CuArrays.CuArrayBackend, ::Function, ::Tuple{CuArray{Tuple{Float32,Zygote.var"#963#back#398"{Zygote.var"#396#397"{Float32}}},1,Nothing},Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64}},Zygote.var"#1666#1673"{Zygote.Context,typeof(log)},Tuple{Base.Broadcast.Extruded{CuArray{Float32,1,Nothing},Tuple{Bool},Tuple{Int64}}}}}, ::Int64; name::String) at gpuarrays.jl:32
gpu_call at gpuarrays.jl:23 [inlined]
#gpu_call#1 at execution.jl:60 [inlined]
gpu_call at execution.jl:46 [inlined]
copyto! at broadcast.jl:63 [inlined]
copyto! at broadcast.jl:864 [inlined]
copy at broadcast.jl:840 [inlined]
materialize at broadcast.jl:820 [inlined]
_broadcast at broadcast.jl:121 [inlined]
adjoint at broadcast.jl:130 [inlined]
_pullback at adjoint.jl:47 [inlined]
adjoint at lib.jl:167 [inlined]
_pullback at adjoint.jl:47 [inlined]
broadcasted at broadcast.jl:1232 [inlined]
_crossentropy at stateless.jl:49 [inlined]
#crossentropy#62 at stateless.jl:67 [inlined]
crossentropy at stateless.jl:67 [inlined]
_pullback(::Zygote.Context, ::typeof(crossentropy), ::CuArray{Float32,1,Nothing}, ::CuArray{Float32,1,Nothing}) at interface2.jl:0
loss at error.jl:13 [inlined]
_pullback(::Zygote.Context, ::typeof(loss), ::CuArray{Float32,1,Nothing}, ::CuArray{Float32,1,Nothing}) at interface2.jl:0
adjoint at lib.jl:167 [inlined]
_pullback at adjoint.jl:47 [inlined]
#15 at train.jl:88 [inlined]
_pullback(::Zygote.Context, ::Flux.Optimise.var"#15#23"{typeof(loss),Tuple{CuArray{Float32,1,Nothing},CuArray{Float32,1,Nothing}}}) at interface2.jl:0
pullback(::Function, ::Zygote.Params) at ...
```

If I use logitcrossentropy instead of crossentropy I do not get an error or a warning.

**Comment by ararslan** [comment]
Can you try with Flux master? Compatibility with CuArrays 2 merged quite recently and is not yet in a release.

**Comment by OTapio** [comment]
I did not have this problem with Flux 0.8.3 and CuArrays 1.7.3.

**Comment by CarloLucibello** [comment]
assume this to be fixed by #1098 on the just released version. If that's not the case, feel free to reopen
